llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeFlashAttentionPass.cpp

```diff
@@ -0,0 +1,466 @@
+        SmallVector<int64_t>(2, ShapedType::kDynamic), elementType);
+    Value qkTranspose =
+        computeQKTranspose(querySlice, keySlice, empty, emptySquare, zeroF32,
+                           tensorType, loc, rewriter);
+
+    empty = rewriter.create<tensor::EmptyOp>(
+        loc, SmallVector<OpFoldResult>{sequenceTileLength}, elementType);
+
+    // Compute current statistics
+    Value currentMax = computeRowwiseReduction<arith::MaxFOp>(
+        qkTranspose, negativeMax, loc, rewriter);
+    Value softmax =
+        computeSoftmax(qkTranspose, currentMax, emptySquare, loc, rewriter);
+    Value currentSum =
+        computeRowwiseReduction<arith::AddFOp>(softmax, zeroSum, loc, rewriter);
+
+    // Update global statistics
+    Value newMax = computeNewMax(iterArgMax, currentMax, empty, loc, rewriter);
+    Value oldWeight =
+        subtractAndExponentiate1D(iterArgMax, newMax, empty, loc, rewriter);
+    Value currentWeight =
+        subtractAndExponentiate1D(currentMax, newMax, empty, loc, rewriter);
+    Value newSum = computeNewSum(iterArgSum, currentSum, oldWeight,
+                                 currentWeight, empty, loc, rewriter);
+
+    // Scale softmax
+    softmax = scaleSoftmax(softmax, currentWeight, newSum, emptySquare, loc,
+                           rewriter);
+
+    // Update accumulator
+    empty = rewriter.create<tensor::EmptyOp>(
+        loc, SmallVector<OpFoldResult>{sequenceLength, headDimension},
+        elementType);
+    Value scaledAcc = scaleAccumulator(outputSlice, iterArgSum, newSum,
+                                       oldWeight, empty, loc, rewriter);
+
+    // Compute matmul(softmax, v)
+    Value result = rewriter
+                       .create<linalg::MatmulOp>(
+                           loc, outputSlice.getType(),
+                           ValueRange{softmax, valueSlice}, scaledAcc)
+                       .getResult(0);
+
+    // Insert slices
+    auto [updatedAcc, updatedMax, updatedSum] = insertSlices(
+        result, iterArgResult, newMax, iterArgMax, newSum, iterArgSum,
+        queryShape, ivs, sequenceTileLength, loc, rewriter);
+
+    if (scf::YieldOp yieldOp = dyn_cast<scf::YieldOp>(
+            secondLoopNest.loops.back().getBody()->getTerminator())) {
+      rewriter.replaceOpWithNewOp<scf::YieldOp>(
+          yieldOp, ValueRange{updatedAcc, updatedMax, updatedSum});
+    }
+
+    if (scf::YieldOp yieldOp = dyn_cast<scf::YieldOp>(
+            firstLoopNest.loops.back().getBody()->getTerminator())) {
+      OpBuilder::InsertionGuard yieldGuard(rewriter);
+      rewriter.setInsertionPoint(yieldOp);
+      rewriter.replaceOpWithNewOp<scf::YieldOp>(
+          yieldOp, ValueRange{secondLoopNest.results[0]});
+    }
+
+    attnOp.getResults()[0].replaceAllUsesWith(firstLoopNest.results[0]);
+    return WalkResult::advance();
+  });
+  return success();
+}
+
+} // namespace
+
+namespace {
+struct TileAndDecomposeFlashAttentionTransformPass
```

Comment:
I'm trying to understand why we cannot tile using the TilingInterface and the existing tileAndDistribute pass. Is there a good place to read about that? I thought the first level of tiling would just tile along a common parallel dimension.

Reply:
If you look at the attention patch https://github.com/iree-org/iree/pull/11928, this get us pretty close to the original flash attention paper. Here, we do an entire row of the first matmul, then the softmax and then the second matmul. The problem is that in general that row can be quite long. So we need to tile that row and do it one tile at a time. This means also maintaining statistics about the other linalg generics. Would be more than happy to go over it in GVC. Let me know what's a good time.

Reply:
Also here is a link to the paper: https://arxiv.org/pdf/2205.14135.pdf

Reply:
@ThomasRaoux might need to use the Reduction tiling interface you added, but with slight changes to how the initialization and merge is used... thats part is to be resolved going forward.

Reply:
I was looking at the PartialReductionOpInterface here (https://github.com/llvm/llvm-project/blob/b41eb9601cf1aa6e33eedbb1c923b8ed0dcdcd42/mlir/include/mlir/Interfaces/TilingInterface.td#L159) but its not obvious how to extend this to multiple ops. (matmuls + linalg generics, with no fixup needed for matmuls but needed for generics)